Core/src/org/sleuthkit/autopsy/geolocation/datamodel/WaypointBuilder.java

```diff
@@ -0,0 +1,490 @@
+
+    /**
+     * Create the query for getting a list of waypoints that do not have time
+     * stamps.
+     *
+     * @param dataSources List of data Sources to filter by
+     *
+     * @return SQL SELECT statement
+     */
+    static private String buildQueryForWaypointsWOTimeStamps(List<DataSource> dataSources) {
+        String query = "";
+
+        query = String.format(SELECT_WO_TIMESTAMP,
+                String.format(GEO_ARTIFACT_QUERY_ID_ONLY,
+                        BlackboardAttribute.ATTRIBUTE_TYPE.TSK_DATETIME.getTypeID(),
+                        BlackboardAttribute.ATTRIBUTE_TYPE.TSK_DATETIME_CREATED.getTypeID()),
+                getWaypointListQuery(dataSources));
+
+        return query;
+    }
+
+    /**
+     * Build the query to filter the list of waypoints.
+     *
+     * If showAll is true, the values of cntDaysFromRecent and noTimeStamp are
+     * ignored.
+     *
+     * @param dataSources       This of data sources to filter the waypoints by.
+     *                          Pass a null or empty list to show way points for
+     *                          all dataSources.
+     *
+     * @param showAll           True to get all waypoints.
+     *
+     * @param cntDaysFromRecent Number of days from the most recent time stamp
+     *                          to get waypoints for. This parameter will be
+     *                          ignored if showAll is true;
+     *
+     * @param noTimeStamp       True to include waypoints without timestamp.
+     *                          This parameter will be ignored if showAll is
+     *                          true.
+     *
+     * @return
+     */
+    static private String buildQuery(List<DataSource> dataSources, boolean showAll, int cntDaysFromRecent, boolean noTimeStamp) {
+        String mostRecentQuery = "";
+
+        if (!showAll && cntDaysFromRecent > 0) {
+            mostRecentQuery = String.format("AND value_int64 > (%s)",  //NON-NLS
+                    String.format(MOST_RECENT_TIME,
+                            cntDaysFromRecent,
+                            BlackboardAttribute.ATTRIBUTE_TYPE.TSK_DATETIME.getTypeID(),
+                            BlackboardAttribute.ATTRIBUTE_TYPE.TSK_DATETIME_CREATED.getTypeID(),
+                            getWaypointListQuery(dataSources)
+                    ));
+        }
+
+        // This givens us all artifact_ID that have time stamp
+        String query = String.format(GEO_ARTIFACT_QUERY,
+                BlackboardAttribute.ATTRIBUTE_TYPE.TSK_DATETIME.getTypeID(),
+                BlackboardAttribute.ATTRIBUTE_TYPE.TSK_DATETIME_CREATED.getTypeID());
+
+        // That are in the list of artifacts for the given data Sources
+        query += String.format("AND artifact_id IN(%s)", getWaypointListQuery(dataSources)); //NON-NLS
+        query += mostRecentQuery;
+
+        if (noTimeStamp) {
+            query = String.format("%s UNION %s", buildQueryForWaypointsWOTimeStamps(dataSources), query); //NON-NLS
```

Comment:
Currently items which do not have a timestamp are not being included when show all is selected unless the no time stamp checkbox is checked, however since the no time stamp checkbox is disabled when show all is selected this results in a user experience that is not desirable.

Reply:
Good catch.  Thank you.